src/KDMqtt/mqtt.cpp

```diff
@@ -0,0 +1,523 @@
+/*
+  This file is part of KDUtils.
+
+  SPDX-FileCopyrightText: 2024 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
+  Author: Marco Thaller <marco.thaller@kdab.com>
+
+  SPDX-License-Identifier: MIT
+
+  Contact KDAB at <info@kdab.com> for commercial licensing options.
+*/
+#include "mqtt.h"
+#include <spdlog/spdlog.h>
+
+namespace KDMqtt {
+
+constexpr std::chrono::milliseconds c_miscTaskInterval = std::chrono::milliseconds(1000);
+
+using namespace KDFoundation;
+
+MqttLib::MqttLib()
+    : m_isInitialized{ false }
+    , m_mosquittoLib(&MosquittoLib::instance())
+{
+}
+
+MqttLib &MqttLib::instance()
+{
+    static MqttLib s_instance;
+    return s_instance;
+}
+
+int MqttLib::init()
+{
+    int result = MOSQ_ERR_UNKNOWN;
+
+    if (!m_isInitialized) {
+        result = m_mosquittoLib->init();
+        const auto hasError = checkMosquittoResultAndDoDebugPrints(result, "MqttLib::init()");
+        m_isInitialized = !hasError;
+        if (m_isInitialized) {
+            int major, minor, revision = 0;
+            version(&major, &minor, &revision);
+            spdlog::info("MqttLib::init() - using libmosquitto v{}.{}.{}", major, minor, revision);
+        }
+    } else {
+        spdlog::warn("MqttLib::init() - Library is already initialized.");
+    }
+    return result;
+}
+
+int MqttLib::cleanup()
+{
+    const auto result = m_mosquittoLib->cleanup();
+    const auto hasError = checkMosquittoResultAndDoDebugPrints(result, "MqttLib::cleanup()");
+    m_isInitialized = hasError ? m_isInitialized : false;
+    return result;
+}
+
+bool MqttLib::isInitialized() const
+{
+    return m_isInitialized;
+}
+
+bool MqttLib::isValidTopicNameForSubscription(const std::string &topic)
+{
+    return m_mosquittoLib->isValidTopicNameForSubscription(topic);
+}
+
+int MqttLib::version(int *major, int *minor, int *revision)
+{
+    return m_mosquittoLib->version(major, minor, revision);
+}
+
+bool MqttLib::checkMosquittoResultAndDoDebugPrints(int result, std::string_view func)
+{
+    const auto isError = (result != MOSQ_ERR_SUCCESS);
+    if (isError) {
+        const auto funcString = func.empty() ? "mosquitto function" : func;
+        spdlog::error("{} - error: {}", funcString, errorString(result));
+    }
+    return isError;
+}
+
+std::string_view MqttLib::connackString(int connackCode)
+{
+    return m_mosquittoLib->connackString(connackCode);
+}
+
+std::string_view MqttLib::errorString(int errorCode)
+{
+    return m_mosquittoLib->errorString(errorCode);
+}
+
+std::string_view MqttLib::reasonString(int reasonCode)
+{
+    return m_mosquittoLib->reasonString(reasonCode);
+}
+
+MqttClient::MqttClient(const std::string &clientId, bool cleanSession, bool verbose)
+    : m_verbose{ verbose }
+{
+    if (!MqttLib::instance().isInitialized()) {
+        spdlog::warn("MqttClient::MqttClient() - CTOR called before MqttLib::init(). Initialize lib before instantiating MqttClient object!");
```

Comment:
That's an error IMO. The class can't really function properly without MqttLib being initialized.